llama_cpp/llama_chat_format.py

```diff
@@ -1686,9 +1687,731 @@ def message_to_str(msg: llama_types.ChatCompletionRequestMessage):
+        else:
+            return prompt, completion_tokens, "content"
+        
+    def generate_content(self, prompt: str, content: str, completion_tokens):
+        stops = [self.tool_call_token, self.stop_token]
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=None)
+        completion_text = completion["choices"][0]["text"]
+        completion_tokens += completion["usage"]["completion_tokens"]
+        prompt += completion_text
+        content_prefix = "<|start_header_id|>assistant<|end_header_id|>\n\n"
+        content += prompt[prompt.rindex(content_prefix) + len(content_prefix):] + completion_text
+        
+        return content, prompt, completion_tokens, completion
+        
+    def generate_tool_call(self, prompt: str, completion_tokens):
+        # Generate function name
+        stops = ["\n"]  # "{self.tool_call_token}{func_name}\n"
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=None)
+        completion_text = completion["choices"][0]["text"]
+        completion_tokens += completion["usage"]["completion_tokens"]
+        function_name = completion_text.rstrip()
+        prompt += completion_text + "\n"
+        # Generate function args
+        stops = [self.tool_call_token, self.stop_token]
+        grammar = self.get_grammar(function_name) if function_name != "python" else None
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=grammar)
+        function_args = completion["choices"][0]["text"]
+        completion_tokens += completion["usage"]["completion_tokens"]
+        prompt += function_args
+
+        return function_name, function_args, prompt, completion_tokens, completion
+    
+    def generate_streaming_to_determine_text_or_tool_call(
+        self, prompt: str, tool_func_choice: Optional[Union[str, Dict]]
+    ):
+        delta_texts = []
+        completion_text = ""
+        # Generate 1 token first to determine text or tool call
+        stops = []
+        grammar = None
+        completion = self.create_completion(
+            prompt=prompt, stop=stops, grammar=grammar, max_tokens=1
+        )
+        for chunk in completion:
+            delta_text = chunk["choices"][0]["text"]
+            if delta_text:
+                delta_texts.append(delta_text)
+                completion_text += delta_text
+        
+        prompt += completion_text
+        
+        if completion_text.strip() == self.tool_call_token:
+            return delta_texts, prompt, "tool_call"
+        elif len(completion_text) == 0:
+            return delta_texts, prompt, "stop"
+        else:
+            return delta_texts, prompt, "content"
+        
+    def generate_streaming_content(self, prompt: str):
+        stops = [self.tool_call_token, self.stop_token]
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=None)
+        for chunk in completion:
+            delta_text = chunk["choices"][0]["text"]
```

Comment:
Does `generate_streaming_content` for v2.5 also need to strip the `content_prefix` assistant message?